gensim/models/coherencemodel.py

```diff
@@ -148,30 +169,84 @@ def __init__(self, model=None, topics=None, texts=None, corpus=None, dictionary=
             self.window_size = window_size
+            if self.window_size is None:
+                self.window_size = SLIDING_WINDOW_SIZES[self.coherence]
             if texts is None:
                 raise ValueError("'texts' should be provided for %s coherence." % coherence)
             else:
                 self.texts = texts
         else:
             raise ValueError("%s coherence is not currently supported." % coherence)
+
         self.topn = topn
-        self.model = model
+        self._model = model
+        self._accumulator = None
+        self._topics = None
+        self.topics = topics
+
+        self.processes = processes if processes > 1 else max(1, mp.cpu_count() - 1)
+
+    def __str__(self):
+        return str(self.measure)
+
+    @property
+    def model(self):
+        return self._model
+
+    @model.setter
+    def model(self, model):
+        self._model = model
         if model is not None:
-            self.topics = self._get_topics()
+            new_topics = self._get_topics()
+            self._update_accumulator(new_topics)
+            self._topics = new_topics
+
+    @property
+    def measure(self):
+        return COHERENCE_MEASURES[self.coherence]
+
+    @property
+    def topics(self):
+        return self._topics
+
+    @topics.setter
+    def topics(self, topics):
+        new_topics = None
+        if self.model is not None:
+            new_topics = self._get_topics()
+            if topics is not None:
+                logger.warning(
+                    "Ignoring topics you are attempting to set in favor of model's topics: %s",
+                    self.model)
         elif topics is not None:
-            self.topics = []
+            new_topics = []
             for topic in topics:
-                t_i = []
-                for n, _ in enumerate(topic):
-                    t_i.append(dictionary.token2id[topic[n]])
-                self.topics.append(np.array(t_i))
-        self.coherence = coherence
+                topic_token_ids = np.array([self.dictionary.token2id[token] for token in topic])
+                new_topics.append(topic_token_ids)
 
-    def __str__(self):
-        return coherence_dict[self.coherence].__str__()
+        self._update_accumulator(new_topics)
+        self._topics = new_topics
+
+    def _update_accumulator(self, new_topics):
+        if self._relevant_ids_will_differ(new_topics):
+            logger.debug("Wiping cached accumulator since it does not contain all relevant ids.")
+            self._accumulator = None
+
+    def _relevant_ids_will_differ(self, new_topics):
+        if self._accumulator is None or not self._topics_differ(new_topics):
+            return False
+
+        new_set = unique_ids_from_segments(self.measure.seg(new_topics))
+        return not self._accumulator.relevant_ids.issuperset(new_set)
+
+    def _topics_differ(self, new_topics):
+        return (new_topics is not None and
+                    self._topics is not None and
```

Comment:
This is, IMO, more confusing if limited to one line. I changed this to the standard hanging indent (8 spaces).